feat: enable Ledger integration

dawnseeker8 opened the pull request:
## **Description**
Previous [PR](https://github.com/MetaMask/metamask-mobile/pull/8128/files#diff-79b4fb7cf0ca87d9c250d6a583b48071c092027de0100aff6b8fecb2a3a44ba4) have removed ledger features and its dependancies in main branch, this PR will re-enable thee ledger feature in those modify files.

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**


1. In `Connect a hareware wallet screen`, tap `Select a hardware wallet you'd like to use` label 7 times to enable ledger feature.
2. tap `ledger` icon to enable bluetooth to import account from ledger
3. Go to [metamask test-dapp](https://metamask.github.io/test-dapp/) to test most of functions still work fine with ledger account.

## **Screenshots/Recordings**
- Connect a hareware wallet screen.
<img width="456" alt="Screenshot 2024-01-11 at 13 23 55" src="https://github.com/MetaMask/metamask-mobile/assets/7315988/f3f23e5e-ab87-4ef1-9771-376584deeec6">

- account dropdown list:
<img width="457" alt="Screenshot 2024-01-11 at 13 27 09" src="https://github.com/MetaMask/metamask-mobile/assets/7315988/40304154-98b8-4bfc-b3eb-3204cd4fa9b4">

- transation
<img width="464" alt="Screenshot 2024-01-11 at 13 28 10" src="https://github.com/MetaMask/metamask-mobile/assets/7315988/07d039bb-5752-4ad8-855d-7018c17ee76d">

- personal sign
<img width="456" alt="Screenshot 2024-01-11 at 13 30 27" src="https://github.com/MetaMask/metamask-mobile/assets/7315988/88862c67-5354-48f9-947c-19f22cc5dffc">

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

dawnseeker8 committed:
feat: Reenable Ledger feature by revert previous PR code.
Also fix the code issue in ledger.ts file.

dawnseeker8 committed:
feat: Fix lint:tsc error in pipeline.

dawnseeker8 committed:
feat: run `yarn deduplicate` to remove duplicate dependancies.

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

mikesposito commented on app/core/Ledger/Ledger.ts:
```diff
@@ -1,60 +1,61 @@
+  return keyringController.getEthKeyringController();
+};
 
+/**
+ * A method replicate #fullupdate from the keyring controller.
+ */
+export const syncKeyringState = async (): Promise<void> => {
+  const keyringController = Engine.context.KeyringController;
+  const ethKeyringController = getEthKeyringController();
+  const { vault } = ethKeyringController.store.getState();
+  const { keyrings, isUnlocked, encryptionKey, encryptionSalt } =
+    ethKeyringController.memStore.getState();
+  keyringController.update(() => ({
+    vault,
+    keyrings,
+    isUnlocked,
+    encryptionKey,
+    encryptionSalt,
+  }));
+};
```
Why do we need to replicate the `#fullUpdate` method manually here? Is there some issue with how `KeyringController` currently does that?

dawnseeker8 replied:
This code has been removed in latest update. this is redundant code before the keyringController v8.1 update. The latest version already fully use KeyringController V8.1 API by removing the EthKeyringController in the ledger.ts

mikesposito commented on app/core/Ledger/Ledger.ts:
```diff
@@ -1,60 +1,61 @@
+ *
+ * @returns The EthKeyringController
+ */
+const getEthKeyringController = () => {
+  const keyringController = Engine.context.KeyringController;
+  return keyringController.getEthKeyringController();
+};
 
```
Do we need some EthKeyringController's method that is not available through KeyringController?

dawnseeker8 replied:
The latest change has completely removed EthKeyringController  and to use keyringController v8.1 directly in the code now.

mikesposito commented on app/core/Ledger/Ledger.ts:
```diff
@@ -96,62 +96,56 @@ export const unlockLedgerDefaultAccount = async (): Promise<{
-  const newAccounts = await keyringController.getAccounts();
-
+  const oldAccounts = await ethKeyringController.getAccounts();
+  await ethKeyringController.addNewAccount(keyring);
+  const newAccounts = await ethKeyringController.getAccounts();
   keyringController.updateIdentities(newAccounts);
   newAccounts.forEach((address: string) => {
     if (!oldAccounts.includes(address)) {
       if (keyringController.setAccountLabel) {
         // The first ledger account is always returned.
         keyringController.setAccountLabel(address, `${keyring.getName()} 1`);
       }
-      preferencesController.setSelectedAddress(address);
+      keyringController.setSelectedAddress(address);
     }
   });
-  await keyringController.persistAllKeyrings();
-
+  await ethKeyringController.persistAllKeyrings();
+  await syncKeyringState();
   const address = await keyring.getDefaultAccount();
   return {
     address,
     balance: `0x0`,
   };
 };
```
Could this be the equivalent of calling `KeyringController.addNewAccountForKeyring`?

https://github.com/MetaMask/core/blob/3f05a94852d96e62c3d90743b2c4dbccf0092972/packages/keyring-controller/src/KeyringController.ts#L431

dawnseeker8 replied:
This method has been modified to remove EthKeyringController in the middle. this comment is not valid anymore,

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

dawnseeker8 committed:
feat: Remove all patch code for keyringController 8.1,
Change Ledger code to remove EthKeyringController,
FIx some issue for TypedSignV4

dawnseeker8 committed:
feat: remove all console code in ledger.ts

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

dawnseeker8 committed:
feat: push updated podfile.lock

dawnseeker8 committed:
feat; fix the lint dequeue error

mikesposito commented on app/core/Ledger/Ledger.ts:
```diff
@@ -44,7 +26,7 @@ export const addLedgerKeyring = async (): Promise<LedgerKeyring> => {
   const keyringController = Engine.context.KeyringController;
   // There should only be one ledger keyring.
-  const keyring = keyringController().getKeyringsByType(
+  const keyring = keyringController.getKeyringsByType(
```
This could also be `undefined`, so probably it would be better to reflect this in the function return type

gantunesr replied:
Yeah, check the method in `eth-keyring-controller`

```ts
getKeyringsByType(type: string): EthKeyring<Json>[] {
  return this.keyrings.filter((keyring) => keyring.type === type);
}
```

It can return an empty array

Akaryatrh replied:
We don't return `undefined` anymore, please check changes here 🙂 https://github.com/MetaMask/metamask-mobile/pull/8246/files#diff-79b4fb7cf0ca87d9c250d6a583b48071c092027de0100aff6b8fecb2a3a44ba4R33

mikesposito commented on app/core/Ledger/Ledger.ts:
```diff
@@ -96,31 +78,18 @@ export const unlockLedgerDefaultAccount = async (): Promise<{
-  });
-  await keyringController.persistAllKeyrings();
-
+  try {
+    await keyringController.addNewAccountForKeyring(keyring);
+  } catch (e) {
+    console.error(e);
+  }
   const address = await keyring.getDefaultAccount();
   return {
     address,
     balance: `0x0`,
   };
```
What happens if the account fails to be added as we ignore the error?

Moreover, `keyringController.addNewAccountForKeyring` returns the added account: could we use that instead of calling `getDefaultAccount`?

dawnseeker8 replied:
First of all, `unlockLedgerDefaultAccount` method has been used in `LedgerConnect` (during bluetooth paired with ledger and then import ledger account and message signing flow), and then `LedgerTransactionConfirmationModal' for all ledger transaction interaction.

Previously, `keyringController.addNewAccountForKeyring()` doesn't throw any exception if keyring already exists in keyringController, but for somehow, after `KeyringController v8.1 upgrade`, `addNewAccountForKeyring()` start to throw exception during transaction flow and signing flow and transaction flow.  the above code has been introduce to quickly fix the issue during `signing` and `transactioni` flow.

However, i agreed that above try catch block is not robust, currently i try to find a way to avoid running `addNewAccountForKeyring()`, i will check whether selectedAccount in metamask is  ledger defaultAccount, if yes, which mean the operation flow is `ledger message signing` or `ledger transaction flow`.  if selectedAccount is not ledger defaultAccount, then, `unlockLedgerDefaultAccount` will be called by pairing ledger and import ledger account. then `addNewAccountForKeyring()` should be run.

@mikesposito To answer your question, i will remove the `try catch` block above, so if during pairing ledger and import ledger account flow failed to add account. then ` await keyringController.addNewAccountForKeyring(keyring);` will directly throw exception for system to catch.

dawnseeker8 replied:
have modified the `unlockLedgerDefaultAccount` method to include extra flag as argument to check whether the call to this method require to import a new account or not.  i have removed the `try catch` block in that method as well so that `addNewAccountForKeyring` can throw exception if account fails to be added.

dawnseeker8 replied:
hi, @mikesposito yea, the exception will bubble up and catch in the ledger connect screen when adding account,  that is previous code did.

gantunesr commented on patches/@metamask+keyring-controller+8.1.0.patch:
```diff
@@ -11,7 +11,7 @@ index 08a8714..a096388 100644
 +     * This patch addresses an issue regarding the forget device functionality. It
 +     * improves the logic  to correctly remove the QR accounts and update the
 +     * identities as needed.
-+     * 
++     *
 +     * Solved in https://github.com/MetaMask/core/pull/3641
 +     * ===============================================================================
 +     */
```
I see that the patch regarding the `forgetQRDevice` is missing, can you make sure to add it back?

gantunesr commented on app/core/Ledger/Ledger.ts:
```diff
@@ -44,7 +26,7 @@ export const addLedgerKeyring = async (): Promise<LedgerKeyring> => {
   // There should only be one ledger keyring.
-  const keyring = keyringController().getKeyringsByType(
+  const keyring = keyringController.getKeyringsByType(
     ExtendedKeyringTypes.ledger,
   )[0] as unknown as LedgerKeyring;
 
```
You can remove the `unknown` type cast since `getKeyringsByType` returns `unknown[]`

dawnseeker8 replied:
unknown type has been removed in latest code.

gantunesr commented on app/core/Ledger/Ledger.ts:
```diff
@@ -1,28 +1,10 @@
-import Engine from '../Engine';
+import LedgerKeyring, {
+  SerializationOptions,
+} from '@consensys/ledgerhq-metamask-keyring';
+import type BleTransport from '@ledgerhq/react-native-hw-transport-ble';
```
We're not doing any tests in this file, since it's a core file inside the client we should make sure to have a 100% of test coverage

Akaryatrh replied:
Unit tests added for the whole core ledger file, please review again! 🙂

Akaryatrh committed:
feat(ledger): add ledger core tests

dawnseeker8 committed:
feat: Improve the unit tests for ledger core.

dawnseeker8 committed:
feat: add unit tests for ledgerSignTypedMessage

dawnseeker8 committed:
feat: Fix a unit tests.

dawnseeker8 committed:
feat: add a flag in UnlockDefaultAccount function in ledger.ts

dawnseeker8 committed:
feat: Remove the console log in the code.

dawnseeker8 committed:
feat: Fix the broken unit tests for unlockDefaultAccount in ledger.ts

Akaryatrh committed:
feat(ledger): add restoreLedgerKeyring to vault + unit tests

Akaryatrh commented on app/core/Vault.js:
```diff
@@ -74,6 +75,9 @@ export const recreateVaultWithNewPassword = async (
     );
   }
 
+  const ledgerKeyring = await getLedgerKeyring();
+  const serializedLedgerKeyring = await ledgerKeyring.serialize();
+
```
Done! Can you review again please 🙂 ?

Akaryatrh committed:
Merge branch 'main' into feat/reenable-ledger-feature

Akaryatrh commented:
I have read the CLA Document and I hereby sign the CLA

dawnseeker8 committed:
feat: Remove ledger beta feature in ledger.

dawnseeker8 committed:
feat: Fix the unit test in ledger.test.ts

dawnseeker8 committed:
feat: try to run `pod update` on ios to see any luck

gantunesr reviewed (commented):
LGTM so far! I need to run a few tests first thing tomorrow and then I'll approve it

dawnseeker8 committed:
feat: revert all transatlon files except english version.

dawnseeker8 committed:
feat: Remove the formatting change because of prettier

dawnseeker8 commented:
I have triggered the Smoke e2e test pipeline , and  regression_e2e_test pipeline in bitrise, and everything are green now.
[Smoke_e2e_test result ](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/31121c03-145a-4960-9b75-4f3ec515042d)
[Pr_regression_e2e_test_result](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c3de7313-807e-4fb7-85e4-f8af3b45d0a2)

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

dawnseeker8 committed:
feat: Add paddingHorizontal and remove height in the LedgerTransactionModal

dawnseeker8 committed:
feat: Update the UI based on Gustova feedback with marginHorizontal space

dawnseeker8 committed:
feat: Fix the button not disabled during loading because isSendingLedgerCommands may be delay to become true.

dawnseeker8 committed:
feat: add one more condition in disabled to prevent some edge case condition during ledger command sending.

gantunesr reviewed (dismissed):
LGTM. Left a small nit comment

dawnseeker8 committed:
feat: Fix the ledger scan click `help` not consistance with QR hardware

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

dawnseeker8 committed:
feat: Prettier format the file to make sure it following the standard.

dawnseeker8 commented on app/core/Ledger/Ledger.ts:
```diff
@@ -88,39 +50,29 @@ export const connectLedgerHardware = async (
   };
 };
-
 /**
```
have just committed the file with prettier format, now the file should have line break in between functions. i dont know why it didnt format since i already turn on `format on save` in my Vs code setting.

mikesposito reviewed (dismissed):
Looks good!

gantunesr reviewed (dismissed):
lgtm!

gantunesr commented:
@SocketSecurity ignore npm/@expo/json-file@8.2.37

gantunesr commented:
![Screenshot 2024-02-06 at 09 41 43](https://github.com/MetaMask/metamask-mobile/assets/17601467/5ef0c69e-e2b9-4f84-bdcb-51c0c0a60eb9)

The new publisher is part of the maintainers.

vivek-consensys commented:
QA passed on Android and iOS using QA build:- https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c3c20b4c-eb95-4a9a-979e-ad62f01fe2b1

Test flows include:- 

- Connect Ledger
- Disconnect Ledger
- Send transaction using Ledger
- Sign typed data v4 using Ledger
- Sign permit using Ledger
- SIWE using Ledger
- Swap using Ledger

dawnseeker8 committed:
Merge branch 'main' into feat/reenable-ledger-feature

dawnseeker8 commented:
Latest e2e pipeline result from bitrise, and they all passed.
1. e2e smoke tests:  https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a9090a22-01bc-4b30-aed4-06b7db39a021
2. regression e2e pipeline: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c14876a1-6cbf-4da1-9628-1f54bacd858a
